### ISSUE TITLE
Cannot expose an API based on Knative service

### ISSUE THREAD

**Issue by abbi-gaurav** [kyma-project/kyma#4922]
**Description**
Since Kyma is built on top of Knative, as a developer I would like to deploy a microservice that is based on Knative service and expose it over the internet via the API leveraging standard security concepts provided by Kyma.
Currently, it is not possible. The API exposed always fails with `no healthy upstream`

**Steps to reproduce**
* Deploy the [Knative sample servcie]. label `serving.knative.dev/visibility: cluster-local` ensures it is not exposed externally by Knative
```yaml
apiVersion: serving.knative.dev/v1alpha1
kind: Service
metadata:
  name: helloworld-go
  labels:
    serving.knative.dev/visibility: cluster-local
spec:
  runLatest:
    configuration:
      revisionTemplate:
        metadata:
          annotations:
            autoscaling.knative.dev/minScale: "1"
        spec:
          container:
            image: gabbi/helloworld-go:latest
            env:
            - name: TARGET
              value: "Go Sample v1"

```
* Expose the helloworld-go service as an API.

**Actual result**

* Accessing the API URL results in `no healthy upstream`

**Expected result**
The exposed API can be accessed without any errors.

**Comment by Abd4llA** [comment]
@abbi-gaurav thanks for reporting it, we are working on adopting Serving and doing seamless integration and exposure of all Serving and Eventing APIs. So this should be fixed in the next few releases. Any urgency from your side for this?

**Comment by piotrmsc** [comment]
Few remarks from my side :  setting label with value cluster-only block incomfing traffic from external world (ingress gateway cannot make request to such pod). if you delete this label, ingress gateway is able to make a call but it's a false positive. Because knative creates a service with type ExternalName pointing to istio ingress! Virtual Service for given external host is pointing to service  in the namespace but since it's an ExternalName kubedn is making DNS redirect to ingress gateway instead of correct k8s service

**Comment by abbi-gaurav** [comment]
Few more notes, 
Knative creates two K8S services for each knative service. e.g. `myservice` and `myservice-<someid>`. It seems the second service has proper cluster ip but the one with proper service name  has the issue pointed out by @piotrmsc .

It will be interesting to know how a developer running Knative services in Kyma is supposed to work with it.

**Comment by rakesh-garimella** [comment]
We should look into following issues here:
1. How can we have internal lambdas if required.
2. How can we expose lambdas securely over internet. Currently knative serving is exposing lambda by creating a virtual service. We should see what should be done with this virtual service or api can leverage this api object to expose lambdas securely.

**Comment by k15r** [comment]
according to my tests:
`serving.knative.dev/visibility: cluster-local ` works fine
```yaml
apiVersion: serving.knative.dev/v1alpha1
kind: Service
metadata:
  labels:
    foo: bar
    serving.knative.dev/visibility: cluster-local
  name: demo-to-destroy
spec:
  template:
    metadata:
      creationTimestamp: null
      name: demo-to-destroy-blubb
    spec:
      containers:
      - env:
        - name: FUNC_HANDLER
          value: main
        - name: MOD_NAME
          value: handler
        - name: FUNC_TIMEOUT
          value: "180"
        - name: FUNC_RUNTIME
          value: nodejs8
        - name: FUNC_MEMORY_LIMIT
          value: 128Mi
        - name: FUNC_PORT
          value: "8080"
        - name: NODE_PATH
          value: $(KUBELESS_INSTALL_VOLUME)/node_modules
        - name: FOO
          value: BAR
        image: k15r/serverless-system-demo-to-destroy:7ed20b6a98f6f0666b83d85b029b27593d3205bb386907f35c30a9d04218d3aa
        name: ""
        resources: {}
      serviceAccountName: function-controller-build
      timeoutSeconds: 300
  traffic:
  - latestRevision: false
    percent: 50
    revisionName: revision-to-destroy-bla
  - latestRevision: false
    percent: 50
    revisionName: revision-to-destroy-blubb
```

setting this label just removes the external gateway from the virtual service. 
```sh
NAME                                         GATEWAYS                                            HOSTS                                                                                                       AGE
route-368cef4f-cef8-11e9-9b8a-42010a8401b9   [kyma-gateway.kyma-system.svc.cluster.local mesh]   [demo-untouched.serverless-system.34.77.25.205.xip.io demo-untouched.serverless-system.svc.cluster.local]   22h
route-608e8f9a-cef8-11e9-9b8a-42010a8401b9   [mesh]                                              [demo-to-destroy.serverless-system.svc.cluster.local]                                                       22h
```
once this is done you can easily create a new virtualservice using the api-controller

this virtual service needs to be modified a bit, so that it can work with knative-serving:

```yaml
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  generation: 1
  labels:
    apiName: demo-to-destroy
    apiNamespace: serverless-system
    apiUid: dcc16cea-cf03-11e9-8aa4-42010a8401f1
    apiVersion: "391043"
    createdBy: api-controller
  name: demo-to-destroy
  selfLink: /apis/networking.istio.io/v1alpha3/namespaces/serverless-system/virtualservices/demo-to-destroy
spec:
  gateways:
  - kyma-gateway.kyma-system.svc.cluster.local
  hosts:
  - demo-to-destroy.34.77.25.205.xip.io
  http:
  # these headers are important!!
  - appendHeaders:
      knative-serving-namespace: serverless-system
      knative-serving-revision: demo-to-destroy-k6gdv
    corsPolicy:
      allowHeaders:
      - Authorization
      - Content-Type
      - '*'
      allowMethods:
      - GET
      - POST
      - PUT
      - DELETE
      allowOrigin:
      - '*'
    match:
    - uri:
        regex: /.*
    route:
    - destination:
        # hostname needs to be a valid revision, so the demo-to-destroy.serverless-system.svc.cluster.local fqdn does not work
        host: demo-to-destroy-k6gdv.serverless-system.svc.cluster.local
        port:
          number: 80
``` 

traffic splitting is also possible using multiple destinations and applied weights:
```yaml
...
    route:
    - destination:
        host: demo-to-destroy-bla.serverless-system.svc.cluster.local
        port:
          number: 80
      weight: 50
    - destination:
        host: demo-to-destroy-blubb.serverless-system.svc.cluster.local
        port:
          number: 80
      weight: 50
```

**Comment by Abd4llA** [comment]
@piotrmsc can you please provide your feedback about the feasibility of needed changes suggested above?

**Comment by piotrmsc** [comment]
@Abd4llA so to recap : 

Once developer will create knative service he needs to : 
- delete created gateway
- create API via API controller 
- add ```  - appendHeaders:
      knative-serving-namespace: default
      knative-serving-revision: demo-to-destroy-bla ``` to created virtual service 

Is that correct? I did not notice any other changes in the VS ?
And my question is why those headers are required?
Value of `knative-serving-namespace` header is the namespace in which API was created?
Value of `knative-serving-revision` header is name os knative svc?

**Comment by k15r** [comment]
`knative-serving-namespace` header ist the namespace in which the knative-service was created.
for the serving-revision, here a knative-serving in a nutshell:
`ksvc` is just a template for the creation of a set of routes, services and pods. With every change to the `ksvc` you get a new revision. the revision header points to one of those.

**Comment by piotrmsc** [comment]
@k15 
ok, from API controller POV we only have info where API CR was created, as long as your controller will create an API in the same namespace as ksvc then it will work 👍 





> `ksvc` is just a template for the creation of a set of routes, services and pods. With every change to the `ksvc` you get a new revision. the revision header points to one of those.

Here you would have to watch for changes in kvs and make updates on API CR respectively as revision is directly used as header.

@Abd4llA From API controller POV there is work to be done with allowing to "specify" the append headers. Question to you is about the priorities for your part, currently, we are developing new API Gateway controller and for me adding this `appendHeaders` to old API Controller has low priority but I'm not saying no :) To speed up the process I'm open for call on slack ofc

**Comment by k15r** [comment]
@piotrmsc imho it would make a lot of sense if the api-controller (old or new) would not only support to configure a service as a target, but also a knativeservice. With this information it could watch the knativeservice, and react on its changes. As the ksvc contains a traffic section in its spec it could reuse this as well for configuring the virtualservice. 
This would also allow all ways how a user can create an API-CR to exist with minimal changes. (e.g UI -> Services -> Expose API, or UI -> API -> Create API)

**Comment by k15r** [comment]
I just verified: the same changes still apply to knative-serving 0.8.1

**Comment by piotrmsc** [comment]
We have two things here :
a) Knative service requires changing something in the gateway created by knative
b) Watching on changes in knative service in the gateway.

The goal of the gateway is to create all required resources for the user in order to expose and secure a service in an easy way. An abstraction from the underlying technology. Use declare wwhat they want : exposed service A with authn/authz, that's all. They do not touch istio or ory,this is controller part.
Controller is creating and managing those sources, when you start messing around with underlying resources it is then your responsibility. From the roadmap perspective, in the future we can think of supporting knative as an uderlying technology but for sure I don't like point a) ito be done in the gateway and I don't like the idea of listening on changes in the knative serivce and then making updates in the underlying resources. For me it should be declarative, I want to update sth in the gateway it is done via gateway CR and changes are populated to underlying resources, from top to bottom, not from bottom to top.

**Comment by antoineco** [comment]
FYI we have a working solution that we will implement this week. See #6496

**Comment by abbi-gaurav** [comment]
@k15r , is this issue still blocked?

**Comment by michal-hudy** [comment]
With the current configuration, it is working. Kservice should has `serving.knative.dev/visibility: cluster-local` label, and then you should create API Rule that exposes lambda, sample API Rule:
```yaml
apiVersion: gateway.kyma-project.io/v1alpha1
kind: APIRule
metadata:
  name: helloworld-go
spec:
  gateway: kyma-gateway.kyma-system.svc.cluster.local
  rules:
  - accessStrategies:
    - config: {}
      handler: noop
    methods:
    - GET
    path: /.*
  service:
    host: helloworld-go.{your_domain_here}
    name: helloworld-go
    port: 80
```
